### PR TITLE
Alpha renaming, yay

### PR DESCRIPTION
**Pull request by andralex**
Now Algebraic can be used for creating self-referenced types. Refer to https://issues.dlang.org/show_bug.cgi?id=9608 for more motivation (probably this ought to close it).

### CODE REVIEW

File: std/typecons.d [hunk 1]
```diff
@@ -6685,3 +6685,213 @@ public:
     auto value = cast(int)flags_A;
     assert(value == Enum.A);
 }
+
+// AlphaRenameType
+/**
+Performs $(LUCKY alpha renaming) of all occurrences of `From` into `To`, in one
+or more types `T`. This is an advanced type manipulation necessary e.g. for
+replacing the placeholder type `This` in $(XREF variant, Algebraic).
```
Comment (WalterBright) [inline std/typecons.d]: Example?

**Comment by jacob-carlborg** [comment]
Issue 9608 is about template introspection.

**Comment by jacob-carlborg** [comment]
In general I would recommend not mixing different type of changes in the same commit. This single commit contains both style changes for comments, replacement of `$(D_PARAM)` to `$(D)` and functionality changes.

It's easier to review if these are separate commits or even separate pull requests.

**Comment by andralex** [comment]
@jacob-carlborg Renaming types was a primary motivator for 9608. Since then there are new traits such as TemplateOf etc. But since there's also https://github.com/D-Programming-Language/dmd/pull/3515 I'll leave it alone.

File: std/typecons.d (hunk 1 above)
Comment (andralex) [inline std/typecons.d]: Will add Returns:. The unittest underneath is the example.

**Comment by andralex** [comment]
@jacob-carlborg yah, variant.d is kinda difficult to review. Let me make a pass trying to break down stuff.

**Comment by andralex** [comment]
Docs:

http://erdani.com/d/phobos-prerelease/std_typecons.html#.AlphaRenameType
http://erdani.com/d/phobos-prerelease/std_variant.html

**Comment by jacob-carlborg** [comment]
Looks a lot less cluttered now with the ddoc macro change as a separate commit.

**Comment by andralex** [comment]
ping

File: std/typecons.d [hunk 2]
```diff
@@ -6685,3 +6685,222 @@ public:
...
+}
+
+private string replaceTypeInFunctionType(X, Y, fun)()
+{
+    alias storageClasses = ParameterStorageClassTuple!fun;
+    string result;
+    static if (functionLinkage!fun == "C")
+    {
+        result ~= "extern(C) ";
+    }
```
Comment (jacob-carlborg) [inline std/typecons.d]: What about other linkages?

File: std/typecons.d [hunk 3]
```diff
@@ -6685,3 +6685,222 @@ public:
...
+    static if (functionAttributes!fun & FunctionAttribute.ref_)
+    {
+        result ~= "ref ";
+    }
+    result ~= (ReplaceType!(X, Y, ReturnType!fun)).stringof;
```
Comment (jacob-carlborg) [inline std/typecons.d]: I've been told that `stringof` shouldn't be used anything other than debugging. Is there something else that can be used?

File: std/typecons.d (hunk 2 above)
Comment (MetaLang) [inline std/typecons.d]: Isn't this more succinctly written as:

    result ~= "extern(" ~ functionLinkage!fun ~ ") ";

Or even:

    result ~= "extern(%s) ".format(functionLinkage!fun);

Comment (andralex) [inline std/typecons.d]: nice

File: std/typecons.d (hunk 3 above)
Comment (andralex) [inline std/typecons.d]: there was no other way I could do it

**Comment by andralex** [comment]
ok, linkage added

File: std/typecons.d (hunk 3 above)
Comment (MetaLang) [inline std/typecons.d]: A lot of the time there really is no other way. A large amount of code in this vein that I've written invariably ends up needing `stringof` as well. Maybe we should provide stronger guarantees about what `stringof` produces.

Comment (andralex) [inline std/typecons.d]: Well it'll break the unittests that's for sure :). The problem here is with the function parameter lists. There is no way to express storage classes such as `ref`, `out`, `lazy` as parts of types.

I also thing `.stringof`for types should offer better guarantees. The simplest one is, for non-Voldemort types, pasting the string into source code should be parsable as a type. Today that's not always the case.

**Comment by andralex** [comment]
ping

**Comment by WalterBright** [comment]
Auto-merge toggled on

**Comment by CyberShadow** [comment]
This pull request introduced a regression:
https://issues.dlang.org/show_bug.cgi?id=15039

File: std/typecons.d [hunk 4]
```diff
@@ -6685,3 +6685,219 @@ public:
...
+{
+    alias storageClasses = ParameterStorageClassTuple!fun;
+    string result;
+    result ~= "extern(" ~ functionLinkage!fun ~ ") ";
+    static if (functionAttributes!fun & FunctionAttribute.ref_)
+    {
+        result ~= "ref ";
+    }
+    result ~= (ReplaceType!(X, Y, ReturnType!fun)).stringof;
+    static if (is(fun == delegate))
```
Comment (MartinNowak) [inline std/typecons.d]: > There is no way to express storage classes such as ref, out, lazy as parts of types.

Which has always been an annoying limitation.
You could alias the type first (alias RT = ReplaceType!(X, Y, ReturnType!fun);) and then use `"ref RT"`.

**Comment by CyberShadow** [comment]
This pull request introduced a regression:
https://issues.dlang.org/show_bug.cgi?id=15168

File: std/typecons.d (hunk 4 above)
Comment (9rnsr) [inline std/typecons.d]: Yes, and I used the method in #3761.

**Pull request merged**